### ISSUE TITLE
STT0083: (Exp 2b) Experimenting on speech to text models for different Tibetan dialects.

### ISSUE THREAD

**Issue by gangagyatso4364** [OpenPecha/stt-wav2vec2#14]
### **Experiment Card**

---

**Title**: Fine-Tuning a Multidialect Speech-to-Text Model for Tibetan (Utsang, Khampa, Amdo)

---

**Objective**:  
To develop a speech-to-text model that accurately transcribes three Tibetan dialects (Utsang, Khampa, Amdo) by exploring different fine-tuning approaches.

---

### **Datasets**  
- **Utsang**: 1500 hours (pretrained base model exists).  
- **Khampa**: 80 hours.  
- **Amdo**: 53 hours.  

---

#### **Experiment 1: Train Separate Models for Each Dialect**  

**Pros:**  
- **Dialect-Specific Optimization:** Each model can be tailored to the nuances of its respective dialect.  
- **Potential for Higher Accuracy:** Specialized models might achieve better performance on their target dialect.  

**Cons:**  
- **Lack of Generalization:** Models trained on limited data may struggle with unseen variations.  
- **Computational Cost:** Requires training three separate models.  
- **Data Scarcity:** Khampa and Amdo datasets are relatively small, which can lead to overfitting.  

---
#### **Experiment 2: Train a Unified Model for All Dialects**  

**2a. Train from Scratch**  

**Pros:**  
- **Unified Model:** A single model can handle all dialects.  
- **Potential for Better Generalization:** Exposure to diverse data can improve robustness.  

**Cons:**  
- **Computational Cost:** Training a large model from scratch is expensive.  
- **Data Imbalance:** The significant disparity in dataset sizes can bias the model towards Utsang.  

**2b. Fine-Tune Existing Utsang Model**  

**Pros:**  
- **Leverages Existing Knowledge:** The pretrained Utsang model provides a strong foundation.  
- **Computational Efficiency:** Requires less training time compared to training from scratch.  

**Cons:**  
- **Catastrophic Forgetting:** The model may forget information about the Utsang dialect during fine-tuning.  
- **Data Imbalance:** Similar to 2a, the imbalance can negatively impact performance.  

---

### **Recommendation**  

**Approach 2b** (Fine-Tune Existing Utsang Model)  
- **Why**: Efficient, leverages strong Utsang base, and adapts well to low-resource dialects.  
- **Solution**: Use balanced sampling to avoid bias toward Utsang and mitigate catastrophic forgetting.  

---

### **Evaluation**  

- **Metrics**: **CER** (Character Error Rate) and **WER** (Word Error Rate).  
- **Success**: Improved transcription for Khampa and Amdo with minimal degradation in Utsang performance.  

---

### Subtask
- [x] extract amdo and kham data from stt pecha tools 
- [x] do cleaning  of audios and transcript and convert to required hf dataset format
- [x] do train validation and test split.
- [x] upload to hf
- [x] do prepare dataset to convert data into proper input for model training
- [x] train the model 
- [x]  upload the best model to hf
- [x] do evaluation of finetuned model and base model on benchmark
- [x] write a documentation of model trained

**Comment by gangagyatso4364** [comment]
we need to allocate gpu for training wa2vec2 on tibetan dialects training data:
training spec:
training duration: one day
training gpu: 4 (L4 24 gb )
Khampa: 80 hours.
Amdo: 53 hours.
epoch : 20

**Comment by gangagyatso4364** [comment]
# Model Evaluation Comparison
### Benchmark stats: Total 8000 samples 2000 from each department
[benchmark](https://huggingface.co/datasets/ganga4364/stt_tibetan_dialects_data/viewer/default/test)

| Department | CER (Fine-Tuned Model) | CER (Base Model) |
|------------|-------------------------|------------------|
| STT_AM     | 0.172302                | 0.442862         |
| STT_AM_AB  | 0.075761                | 0.354756         |
| STT_KH     | 0.208619                | 0.367282         |
| STT_KH_AB  | 0.104608                | 0.310975         |
| **Mean**   | **0.140322**            | **0.368969**     |

**Comment by gangagyatso4364** [comment]
| Model Name              | Model Type | BENCHMARK | Benchmark mean CER | STT_AB  | STT_CS  | STT_MV  | STT_NS  | STT_TT  | STT_HS  | STT_NW  | STT_PC  |
|-------------------------|------------|-----------|--------------------|---------|---------|---------|---------|---------|---------|---------|---------|
| ganga/mms_300_v4        | wav2vec2   | [benchmark v2](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark) | 15.23%             | 8.41%   | 29.89%  | 19.34%  | 10.27%  | 13.24%  | 19.28%  | 11.18%  | 10.23%  |
| Fine-Tuned Model   | wav2vec2   | [benchmark v2](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark)     | **29.62%**         | 19.96%  | 46.79%  | 38.66%  | 25.79%  | 26.72%  | 29.35%  | 22.18%  | 21.24%  |

**Comment by gangagyatso4364** [comment]
# Training arguments
    training_args = TrainingArguments(
        output_dir="/data/volume/dialect_models/",
        per_device_train_batch_size=8,
        gradient_accumulation_steps=2,
        evaluation_strategy="steps",
        save_steps=1000,
        eval_steps=1000,
        logging_steps=100,
        learning_rate=3e-5,
        num_train_epochs=20,
        save_total_limit=100,
        fp16=True,
        warmup_steps=500,
        report_to=['wandb'],
        push_to_hub=False,
        dataloader_num_workers=4
    )

**Comment by gangagyatso4364** [comment]
[blog post](https://forum.openpecha.org/t/fine-tuning-a-multi-dialect-speech-recognition-model-for-tibetan-languages/180)
[wandb](https://wandb.ai/TMLM/stt_finetune/runs/huvlhj9n?nw=nwuserganga2000)
